platform/macos/joypad_macos.mm

```diff
@@ -30,303 +30,275 @@
-		return;
+	_ALWAYS_INLINE_ bool has_active_players() {
+		if (!has_motors()) {
+			return false;
+		}
+		return (weak_motor && weak_motor->has_active_player()) && (strong_motor && strong_motor->has_active_player());
 	}
 
-	// If there is active vibration players, stop them.
-	if ([p_joypad.rumble_context hasActivePlayers]) {
-		joypad_vibration_stop(p_joypad, p_timestamp);
+	void stop() {
+		if (weak_motor) {
+			weak_motor->stop();
+		}
+		if (strong_motor) {
+			strong_motor->stop();
+		}
 	}
 
-	// Gets the default vibration pattern and creates a player for each motor.
-	CHHapticPattern *weak_pattern = get_vibration_pattern(p_weak_magnitude, p_duration);
-	CHHapticPattern *strong_pattern = get_vibration_pattern(p_strong_magnitude, p_duration);
-
-	RumbleMotor *weak_motor = p_joypad.rumble_context.weak_motor;
-	RumbleMotor *strong_motor = p_joypad.rumble_context.strong_motor;
-
-	[weak_motor execute_pattern:weak_pattern];
-	[strong_motor execute_pattern:strong_pattern];
-
-	p_joypad.ff_effect_timestamp = p_timestamp;
-}
+	void play_weak_pattern(CHHapticPattern *p_pattern) {
+		if (weak_motor) {
+			weak_motor->execute_pattern(p_pattern);
+		}
+	}
 
-void JoypadMacOS::joypad_vibration_stop(Joypad *p_joypad, uint64_t p_timestamp) {
-	if (!p_joypad.force_feedback) {
-		return;
+	void play_strong_pattern(CHHapticPattern *p_pattern) {
+		if (strong_motor) {
+			strong_motor->execute_pattern(p_pattern);
+		}
 	}
-	// If there is no active vibration players, exit.
-	if (![p_joypad.rumble_context hasActivePlayers]) {
-		return;
+};
+
+GameController::GameController(int p_joy_id, GCController *p_controller) :
+		joy_id(p_joy_id), controller(p_controller) {
+	force_feedback = NO;
+	if (@available(macOS 11.0, iOS 14.0, tvOS 14.0, *)) {
+		if (controller.haptics != nil) {
+			// Create a rumble context for the controller.
+			rumble_context = memnew(RumbleContext(p_controller));
+
+			// If the rumble motors aren't available, disable force feedback.
+			force_feedback = rumble_context->has_motors();
+		}
 	}
 
-	RumbleMotor *weak_motor = p_joypad.rumble_context.weak_motor;
-	RumbleMotor *strong_motor = p_joypad.rumble_context.strong_motor;
+	int l_joy_id = joy_id;
 
-	[weak_motor stop];
-	[strong_motor stop];
-
-	p_joypad.ff_effect_timestamp = p_timestamp;
-}
+	auto BUTTON = [l_joy_id](JoyButton p_button) {
+		return ^(GCControllerButtonInput *button, float value, BOOL pressed) {
+			Input::get_singleton()->joy_button(l_joy_id, p_button, pressed);
+		};
+	};
 
-@interface JoypadMacOSObserver ()
+	if (controller.extendedGamepad != nil) {
+		GCExtendedGamepad *gamepad = controller.extendedGamepad;
+
+		gamepad.buttonA.pressedChangedHandler = BUTTON(JoyButton::A);
+		gamepad.buttonB.pressedChangedHandler = BUTTON(JoyButton::B);
+		gamepad.buttonX.pressedChangedHandler = BUTTON(JoyButton::X);
+		gamepad.buttonY.pressedChangedHandler = BUTTON(JoyButton::Y);
```

Comment:
Refactored to use the pressedChangedHandler for button-like controls, so that we reduce the number of unnecessary events sent to the Godot input system.

Using a valueChangedHandler meant the buttons would receive continuous events as the pressure changed, where as for buttons, Godot only cares whether it is pressed or not.